---
title: "load_samples"
author: "Camino RSM"
date: "2024-11-06"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown
This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.
When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks.

```{r cars}
setwd("/home/caminors/Desktop/sc_RNAseq")
directory= "/home/caminors/Desktop/sc_RNAseq"
```

### GRUEL ET AL. ###
## Create list of Raw data samples (Gruel et al., 2024) and:
  - Rename cell barcode to biosample_barcode
  - Rename Dimnames from layer@counts to cell_barcode and gene_symbol

```{r, echo=FALSE}
files <- list.dirs(directory, full.names = TRUE)
files <- files[-1]
# Initialize an empty list to store Seurat objects
seurat_object <- list()
# Loop through each folder to read in the 10X data and rename cell barcodes
for (folder in files) {
  # Extract the folder name (e.g., "GSM223" from the path)
  folder_name <- sub("^/mnt/c/Users/caminorsm/Desktop/outs/", "", folder)
  print(paste0("Reading folder: ", folder))
  
  # Read 10X data from the folder
  data <- Seurat::Read10X(folder)
  # Modify cell barcodes by adding the folder name as a prefix
  colnames(data) <- paste0(folder_name, "_", colnames(data))
  seurat = Seurat::CreateSeuratObject(data, project = folder_name)
  seurat@assays[["RNA"]]@layers[["counts"]]@Dimnames <- list(rownames(data), colnames(data))
  # Store the modified matrix in the list with the folder name as the key
  seurat_object[[folder_name]] <- seurat
}
#save list
saveRDS(seurat_object, "./Gruel_list.rds")
```

# Create Seurat object
```{r}
seurat_object = readRDS("/home/caminors/Desktop/sc_RNAseq/Gruel_list.rds")
# merge all seurats into a single seurat
Gruel <- seurat_object[[1]]
# Loop through the rest of the Seurat objects and merge them into 'Gruel'
for (i in 2:length(seurat_object)) {
  print(paste0("Adding: ", names(seurat_object)[[i]]))
  Gruel <- merge(Gruel, y = seurat_object[[i]])
}
#Remove .SeuratProjects from the count.names
names(Gruel@assays$RNA@layers) = gsub("\\.SeuratProject", "", names(Gruel@assays$RNA@layers))
saveRDS(Gruel, "./Gruel.rds")
```
### PDSC ###
## Load CR outputs and cellbarcode renaming
  - Rename cell barcode to biosample_barcode
  - Rename Dimnames from layer@counts to cell_barcode and gene_symbol
```{r}
hsc_files <- files[grep("^hSC", basename(files))]
hsc_files= hsc_files[-1]
for (folder in hsc_files) {
  # Extract the folder name (e.g., "GSM223" from the path)
  folder_name <- sub("^/mnt/c/Users/caminorsm/Desktop/outs/", "", folder)
  print(paste0("Reading folder: ", folder))
  
  # Read 10X data from the folder
  data <- Seurat::Read10X(folder)
  # Modify cell barcodes by adding the folder name as a prefix
  colnames(data) <- paste0(folder_name, "_", colnames(data))
  seurat = Seurat::CreateSeuratObject(data, project = folder_name)
  seurat@assays[["RNA"]]@layers[["counts"]]@Dimnames <- list(rownames(data), colnames(data))
  # Store the modified matrix in the list with the folder name as the key
  seurat_object[[folder_name]] <- seurat
}
saveRDS(seurat_object, "./PDSC_list.rds")
```

## Create Seurat
```{r}
seurat_object = readRDS("/home/caminors/Desktop/sc_RNAseq/PDSC_list.rds")
# merge all seurats into a single seurat
seurat_object[[1]]$orig.ident = names(seurat_object)[[1]] #rename orig.ident correctly
PDSC <- seurat_object[[1]]
# Loop through the rest of the Seurat objects and merge them into 'PDSC'
for (i in 2:length(seurat_object)) {
  seurat_object[[i]]$orig.ident = names(seurat_object)[[i]]
  print(paste0("Adding: ", names(seurat_object)[[i]]))
  PDSC <- merge(PDSC, y = seurat_object[[i]])
}
names(PDSC@assays$RNA@layers) = gsub("\\.SeuratProject", "", names(PDSC@assays$RNA@layers)) # Remove SeuratProject from layer names
saveRDS(PDSC, "/home/caminors/Desktop/sc_RNAseq/Raw_seurats/PDSC.rds")
```
### hSC26 ###
## Load CR outputs and cellbarcode renaming
  - Rename cell barcode to biosample_barcode
  - Rename Dimnames from layer@counts to cell_barcode and gene_symbol
```{r}
  # Read 10X data from the hSC26 folder
  data <- Seurat::Read10X("/home/caminors/Desktop/sc_RNAseq/hSC26/")
  # Create RNA and HTO Assay
  hSC26 = CreateSeuratObject(counts=data$`Gene Expression`)
  hSC26[["HTO"]] = CreateAssay5Object(counts = data$`Antibody Capture`)
  hSC26@assays[["RNA"]]@layers[["counts"]]@Dimnames <- list(rownames(data$`Gene Expression`), colnames(data$`Gene Expression`))
  hSC26@assays[["HTO"]]@layers[["counts"]]@Dimnames <- list(rownames(data$`Antibody Capture`), colnames(data$`Antibody Capture`))
  # Store the modified matrix in the list with the folder name as the key
  saveRDS(hSC26, "Raw_seurats/hSC26.rds")
```

# HTO demultiplexing #
```{r}
library(Seurat)
hSC26 = readRDS("/home/caminors/Desktop/sc_RNAseq/Raw_seurats/hSC26.rds")
# Normalize RNA data with log normalization
hSC26_n <- Seurat::NormalizeData(hSC26, assay = "RNA")
# Find and scale variable features
hSC26_n <- Seurat::FindVariableFeatures(hSC26_n, selection.method = "mean.var.plot")
hSC26_n <- Seurat::ScaleData(hSC26_n, features = VariableFeatures(hSC26_n))
# Normalize HTO data, here we use centered log-ratio (CLR) transformation
Seurat::DefaultAssay(hSC26_n)="HTO"
hSC26_n <- Seurat::NormalizeData(hSC26_n, assay = "HTO", normalization.method = "CLR", margin=2) # set margin=2 due to HTO has to be normalized relatively to other cells rather to the features
hSC26_n <- Seurat::HTODemux(hSC26_n, assay = "HTO", positive.quantile = 0.95)
# Plot the CLR-transformed HTO signal levels
RidgePlot(hSC26_n, assay = "HTO", features = rownames(hSC26_n[["HTO"]]), ncol = 2)
Idents(hSC26_n)="hash.ID"
VlnPlot(hSC26_n, features = c("ddLPS","AdjNorm"))
FeatureScatter(hSC26_n, feature1 = "ddLPS", feature2 = "AdjNorm", slot = "counts")
```
## Filtering doublets and negatives
```{r}
hSC26@meta.data = hSC26_n@meta.data
hSC26_f = subset(hSC26, hash.ID %in% c("Doublet","Negative"),invert=T)
hSC26_f@assays[["HTO"]]@layers[["counts"]]@Dimnames[[1]]=rownames(hSC26_f@assays$HTO)
hSC26_f@assays[["RNA"]]@layers[["counts"]]@Dimnames[[1]]=rownames(hSC26_f@assays$RNA)
## Add this information to orig.ident
hSC26_f$orig.ident = "hSC26"
hSC26_f$orig.ident = paste0(hSC26_f$orig.ident,"_", hSC26_f$hash.ID)
# Combine orig.ident with the original cell barcode to create new cell names
new_cell_names <- paste0(hSC26_f$orig.ident, "_", colnames(hSC26_f))
# Update cell names in the Seurat object
colnames(hSC26_f) <- new_cell_names
rownames(hSC26_f@meta.data) <- new_cell_names
hSC26_f@assays[["RNA"]]@layers[["counts"]]@Dimnames[[2]]=colnames(hSC26_f@assays$RNA)
hSC26_f@assays[["HTO"]]@layers[["counts"]]@Dimnames[[2]]=colnames(hSC26_f@assays$HTO)
names(hSC26_f@assays[["RNA"]]@layers) = "counts"
```

### Merge hSC26 with all PDSC and call it SARC ###

```{r}
PDSC = readRDS("/home/caminors/Desktop/sc_RNAseq/Raw_seurats/PDSC.rds")
SARC = merge(PDSC, hSC26_f)
names(SARC@assays$RNA@layers) = gsub("\\.1", "", names(SARC@assays$RNA@layers)) # rename layer
saveRDS(SARC, "/home/caminors/Desktop/sc_RNAseq/Raw_seurats/SARC_demultiplexed.rds")
```

######  INITAL QUALITY CONTROL ###### 
# 1) Bad quality cell filtering on UMI, n_genes by 0.01 and 0.99 quantiles and MT percentage of 20%
# 2) Removal of ambient RNA contamination using SoupX (based on: "https://github.com/mousepixels/sanbomics_scripts/blob/main/soupX/soupX_R_tutorial.Rmd")
# 3) doublet removal


########################################################################### 

# 1) Removal of low quality cells
```{r}
Gruel= readRDS("/home/caminors/Desktop/sc_RNAseq/Raw_seurats/Gruel.rds")
SARC= readRDS("/home/caminors/Desktop/sc_RNAseq/Raw_seurats/SARC_demultiplexed.rds")
DefaultAssay(Gruel)="RNA"
DefaultAssay(SARC)="RNA"

# plotting
SARC@meta.data %>% 
    ggplot(aes (x=SARC$nCount_RNA, fill= SARC$orig.ident)) + 
    geom_density(alpha = 0.2) + 
    scale_x_log10() + 
    theme_classic() +
    ylab("Cell density") +
    geom_vline(xintercept = 500)
SARC@meta.data %>% 
    ggplot(aes (x=SARC$nFeature_RNA, fill= SARC$orig.ident)) + 
    geom_density(alpha = 0.2) + 
    scale_x_log10() + 
    theme_classic() +
    ylab("Cell density") +
    geom_vline(xintercept = 500)
Gruel@meta.data %>% 
    ggplot(aes (x=Gruel$nCount_RNA, fill= Gruel$orig.ident)) + 
    geom_density(alpha = 0.2) + 
    scale_x_log10() + 
    theme_classic() +
    ylab("Cell density") +
    geom_vline(xintercept = 500)
Gruel@meta.data %>% 
    ggplot(aes (x=Gruel$nFeature_RNA, fill= Gruel$orig.ident)) + 
    geom_density(alpha = 0.2) + 
    scale_x_log10() + 
    theme_classic() +
    ylab("Cell density") +
    geom_vline(xintercept = 500)
```
```{r}
SARC@meta.data  %>% 
  	ggplot(aes(x=nCount_RNA, y=nFeature_RNA, color=percent.mt)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	stat_smooth(method=lm) +
  	scale_x_log10() + 
  	scale_y_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 500) +
  	geom_hline(yintercept = 300) +
  	facet_wrap(~orig.ident)
```

## Detemine 0.01 and 0.99 quantile outliers in Gruel dataset
```{r}
library(dplyr)
library(ggplot2)

# Extract metadata from Seurat object
meta_data <- Gruel@meta.data

# Calculate quantiles and classify cells
meta_data <- meta_data %>%
  group_by(orig.ident) %>%
  mutate(
    nCount_status = case_when(
      nCount_RNA > quantile(nCount_RNA, 0.99) ~ "Upper Outlier",
      nCount_RNA < quantile(nCount_RNA, 0.01) ~ "Lower Outlier",
      TRUE ~ "Within Range"
    ),
    nFeature_status = case_when(
      nFeature_RNA > quantile(nFeature_RNA, 0.99) ~ "Upper Outlier",
      nFeature_RNA < quantile(nFeature_RNA, 0.01) ~ "Lower Outlier",
      TRUE ~ "Within Range"
    )
  ) %>%
  ungroup()

Gruel@meta.data$nCount_status = meta_data$nCount_status
Gruel@meta.data$nFeature_status = meta_data$nFeature_status
table(Gruel$nCount_status)
table(Gruel$nFeature_status)

# Violin plot for nCount_RNA with color coding for outliers
Seurat::VlnPlot(Gruel, features = c("nCount_RNA"), group.by = "orig.ident", fill.by = "nCount_status", pt.size=0) +
  geom_jitter(aes(color = Gruel@meta.data$nCount_status), size = 0.5, width = 0.2) +
  scale_color_manual(values = c("Upper Outlier" = "red", "Lower Outlier" = "blue", "Within Range" = "black")) +
  labs(title = "nCount_RNA Distribution by orig.ident with Outlier Highlights")
# Violin plot for nFeature_RNA with color coding for outliers
VlnPlot(Gruel, features = c("nFeature_RNA"), group.by = "orig.ident", pt.size = 0) +
  geom_jitter(aes(color = Gruel@meta.data$nFeature_status), size = 0.5, width = 0.2) +
  scale_color_manual(values = c("Upper Outlier" = "red", "Lower Outlier" = "blue", "Within Range" = "black")) +
  labs(title = "nFeature_RNA Distribution by orig.ident with Outlier Highlights")

```
## Detemine 0.01 and 0.99 quantile outliers in SARC dataset
```{r}
library(dplyr)

# Extract metadata from Seurat object
meta_data <- SARC@meta.data

# Calculate quantiles and classify cells
meta_data <- meta_data %>%
  group_by(orig.ident) %>%
  mutate(
    nCount_status = case_when(
      nCount_RNA > quantile(nCount_RNA, 0.99) ~ "Upper Outlier",
      nCount_RNA < quantile(nCount_RNA, 0.01) ~ "Lower Outlier",
      TRUE ~ "Within Range"
    ),
    nFeature_status = case_when(
      nFeature_RNA > quantile(nFeature_RNA, 0.99) ~ "Upper Outlier",
      nFeature_RNA < quantile(nFeature_RNA, 0.01) ~ "Lower Outlier",
      TRUE ~ "Within Range"
    )
  ) %>%
  ungroup()

SARC@meta.data$nCount_status = meta_data$nCount_status
SARC@meta.data$nFeature_status = meta_data$nFeature_status
table(SARC$nCount_status)
table(SARC$nFeature_status)

# Violin plot for nCount_RNA with color coding for outliers
Seurat::VlnPlot(SARC, features = c("nCount_RNA"), group.by = "orig.ident", fill.by = "nCount_status", pt.size=0) +
  geom_jitter(aes(color = SARC@meta.data$nCount_status), size = 0.5, width = 0.2) +
  scale_color_manual(values = c("Upper Outlier" = "red", "Lower Outlier" = "blue", "Within Range" = "black")) +
  labs(title = "nCount Distribution by orig.ident with Outlier Highlights")
# Violin plot for nFeature_RNA with color coding for outliers
VlnPlot(SARC, features = c("nFeature_RNA"), group.by = "orig.ident", pt.size = 0) +
  geom_jitter(aes(color = SARC@meta.data$nFeature_status), size = 0.5, width = 0.2) +
  scale_color_manual(values = c("Upper Outlier" = "red", "Lower Outlier" = "blue", "Within Range" = "black")) +
  labs(title = "nFeature Distribution by orig.ident with Outlier Highlights")

```

```{r}
SARC@meta.data %>%
    ggplot(aes(x = nCount_RNA, y = nFeature_RNA, color = nFeature_status)) + 
    geom_point() + 
    scale_colour_manual(values = c("Upper Outlier" = "red", 
                                   "Lower Outlier" = "blue", 
                                   "Within Range" = "grey")) + 
    stat_smooth(method = "lm") +
    scale_x_log10() + 
    scale_y_log10() + 
    theme_classic() +
    geom_vline(xintercept = 500) +
    geom_hline(yintercept = 300) +
    facet_wrap(~orig.ident)
```
# Remove outliers
#  1, Mt
```{r}
Gruel[["percent.mt"]] <- PercentageFeatureSet(Gruel, pattern = "^MT-", assay = "RNA")
SARC[["percent.mt"]] <- PercentageFeatureSet(SARC, pattern = "^MT-", assay = "RNA")
VlnPlot(Gruel, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(SARC, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, group.by = "orig.ident")
```
## Filtering
```{r}
Gruel_f <- subset(Gruel, subset = nCount_status == "Within Range" & nFeature_RNA >300 & nFeature_status =="Within Range" & percent.mt < 20) ## setting these thresholds as they seem the best for SARCOMA dataset :) most nFeature_RNA <300 are actually not "within Range" but just in case
SARC_f <- subset(SARC, subset = nCount_status == "Within Range" & nFeature_RNA >300 & nFeature_status =="Within Range" & percent.mt < 20)

VlnPlot(Gruel_f, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(SARC_f, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, group.by="orig.ident")
```
# Recheck histograms of no-filtered and filtered objects
```{r}
SARC_f@meta.data %>% 
    ggplot(aes (x=SARC_f$nCount_RNA, fill= SARC_f$orig.ident)) + 
    geom_density(alpha = 0.2) + 
    scale_x_log10() + 
    theme_classic() +
    ylab("Cell density") +
    geom_vline(xintercept = 500)
SARC_f@meta.data %>% 
    ggplot(aes (x=SARC_f$nFeature_RNA, fill= SARC_f$orig.ident)) + 
    geom_density(alpha = 0.2) + 
    scale_x_log10() + 
    theme_classic() +
    ylab("Cell density") +
    geom_vline(xintercept = 500)
Gruel_f@meta.data %>% 
    ggplot(aes (x=Gruel_f$nCount_RNA, fill= Gruel_f$orig.ident)) + 
    geom_density(alpha = 0.2) + 
    scale_x_log10() + 
    theme_classic() +
    ylab("Cell density") +
    geom_vline(xintercept = 500)
Gruel_f@meta.data %>% 
    ggplot(aes (x=Gruel_f$nFeature_RNA, fill= Gruel_f$orig.ident)) + 
    geom_density(alpha = 0.2) + 
    scale_x_log10() + 
    theme_classic() +
    ylab("Cell density") +
    geom_vline(xintercept = 500)

```

```{r}
SARC_f@meta.data  %>% 
  	ggplot(aes(x=nCount_RNA, y=nFeature_RNA, color=percent.mt)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	stat_smooth(method=lm) +
  	scale_x_log10() + 
  	scale_y_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 500) +
  	geom_hline(yintercept = 300) +
  	facet_wrap(~orig.ident)
```
```{r}
saveRDS(Gruel_f, "/home/caminors/Desktop/sc_RNAseq/QC_seurats/Gruel_f.rds")
saveRDS(SARC_f, "/home/caminors/Desktop/sc_RNAseq/QC_seurats/SARC_f.rds")
```

### 2) Removal of ambient RNA using SoupX
```{r}
# to clean up the data by identifying and correcting for RNA contamination. These methods typically rely on distinguishing the actual cell expression profiles from background noise (ambient RNA) that may arise from dead cells, empty droplets, or cross-contamination.

library(SoupX)
library(Seurat)
# SoupX requires the raw_feature_bc_matrix from the CellRanger output (output with empty droplets!)
# Since we have renamed the barcodes in the previous steps, we will need to rename the barcodes in this raw_feature_bc_matrix

make_soup <- function(sobj) {
  sample_id = unique(sobj$orig.ident)
  for (sample in sample_id ){
    print(paste0("Reading ", sample))
    path <- paste0("/home/caminors/Desktop/sc_RNAseq/RAW_featured_bc_matrix/outs_raw/", sample)
    raw <- Read10X(data.dir = path)
    if (grepl("hSC26", sample)) {
      sobj_subset = subset(sobj, orig.ident %in% c("hSC26_AdjNorm","hSC26_ddLPS"))
      cell_names = as.data.frame(colnames(sobj_subset))
      print(paste0("Renaming barcodes"))
      colnames(cell_names)="cell_names"
      split_result= strsplit(cell_names$cell_names, "_")
      split_result <- sapply(split_result, `[`, 1)
      cell_names$barcode <- sapply(split_result, `[`, 3)
      raw = raw$`Gene Expression`
      new_colnames <- cell_names$cell_names[match(colnames(raw), cell_names$barcode)]
      colnames(raw)<- new_colnames
      # Remove cells not present in the filtered seurat object
      valid_cols <- !is.na(colnames(raw)) & colnames(raw) != "NA"
      raw_clean <- raw[, valid_cols]
      print(paste0("Seurat object contains: ", ncol(sobj_subset), "cells. Raw matrix contains", ncol(raw_clean), "cells."))
      common_genes = intersect(rownames(raw), rownames(sobj_subset@assays$RNA))
      raw_clean <- raw_clean[common_genes, , drop = FALSE]
    } else {
      sobj_subset= subset(sobj, orig.ident %in% sample)
      print(paste0("Renaming barcodes"))
      colnames(raw) = paste0(sample,"_", colnames(raw))
      # Remove cells not present in the filtered seurat object
      valid_cols <- colnames(sobj_subset)
      raw_clean <- raw[, valid_cols]
      print(paste0("Seurat object contains: ", ncol(sobj_subset), " cells. Raw matrix contains: ", ncol(raw_clean), " cells."))
      common_genes = intersect(rownames(raw), rownames(sobj_subset@assays$RNA))
      raw_clean <- raw_clean[common_genes, , drop = FALSE]
      
      print(paste0("Seurat object contains: ", nrow(sobj_subset), " genes. Raw matrix contains: ", nrow(raw_clean), " genes."))
    }
  sc = SoupChannel(raw_clean,sobj_subset@assays[[,1]])
  sc = setClusters(sc,sobj_subset$soup_group)
  sc = autoEstCont(sc, doPlot=FALSE)
  out = adjustCounts(sc, roundToInt = TRUE)
  seurat = merge(seurat, out)
  #optional keep original
  }
    sobj[["original.counts"]] <- CreateAssayObject(counts = seurat@assays$RNA@layer)
    return(sobj)
}

Gruel_f = make_soup(Gruel_f)

```


